### ISSUE TITLE
04-05-06-Cuadernos-diabetes

### ISSUE THREAD

**Issue by mezcolantriz** [4GeeksAcademy/machine-learning-elius123ef#4]
@elius123ef 

---

Sobre 04-Decission-Tree:

**1. ERROR - Guardado incorrecto del modelo:**

Guardas los datos: pickle.dump((X_train, X_test, y_train, y_test), f)
Deberías guardar el modelo: pickle.dump(dt_model, f) o pickle.dump(dt_model_otro, f)

---

# EJERCICIO 05: RANDOM FOREST

## LO QUE HAS HECHO BIEN

- **Carga elegante de datos anteriores**: Muy buena práctica cargar el modelo y datos del ejercicio anterior desde un pickle. Esto mantiene la continuidad entre ejercicios.

- **Visualización de árboles individuales**: Buena idea mostrar 4 árboles del bosque para entender su estructura.

- **GridSearchCV completo**: Exploras múltiples hiperparámetros (n_estimators, max_depth, min_samples_split, min_samples_leaf, max_features).

- **RandomizedSearchCV adicional**: Excelente que pruebes ambos métodos de búsqueda para comparar.

- **Classification report**: Muy bien incluir el reporte de clasificación con precision, recall y f1.

- **Guardas el bundle completo**: La idea de guardar modelo + datos en un diccionario es muy profesional y facilita el siguiente ejercicio (boosting).

---

### 1. ERROR: class_names con 3 clases cuando solo hay 2

```python
# Tu código:
tree.plot_tree(model.estimators_[0], ..., class_names=["0", "1", "2"], filled=True)
```

El dataset de diabetes tiene **solo 2 clases** (0: no diabético, 1: diabético), pero pones 3 clases.

**Corrección:**
```python
tree.plot_tree(model.estimators_[0], ..., class_names=["No Diabetes", "Diabetes"], filled=True)
```

---

### 2. INFO: Refit innecesario

Creas el modelo manualmente después de obtener los mejores parámetros, pero luego lo sobrescribes:

```python
# Tu código:
best_model = RandomForestClassifier(random_state=42, n_estimators=200, max_depth=None, ...)  # Manual
best_model.fit(X_train, y_train)  # Entrenas

best_model = grid_search.best_estimator_  # ← Lo sobrescribes!
```

El `grid_search.best_estimator_` ya está entrenado. No necesitas ni crear el modelo manualmente ni hacer fit de nuevo:

**Corrección:**
```python
# Después del GridSearchCV.fit(), simplemente:
best_model = grid_search.best_estimator_
y_pred = best_model.predict(X_test)
```

---

### 3. Falta accuracy en train para detectar overfitting

Solo calculas accuracy en test. Deberías comparar train vs test:

```python
print(f"Accuracy Train: {best_model.score(X_train, y_train):.4f}")
print(f"Accuracy Test: {best_model.score(X_test, y_test):.4f}")
```

---

### 4. Falta matriz de confusión

Para un problema médico es importante ver los falsos negativos:

```python
from sklearn.metrics import ConfusionMatrixDisplay

ConfusionMatrixDisplay.from_estimator(best_model, X_test, y_test, 
                                       display_labels=['No Diabetes', 'Diabetes'])
plt.title('Matriz de Confusión - Random Forest')
plt.show()
```
---

# 06. BOOSTING

### LO QUE HAS HECHO BIEN

- **Carga elegante del modelo anterior**: Muy buena práctica guardar un "bundle" con el modelo, X_train, X_test, y_train, y_test en un solo pickle. Esto facilita la continuidad entre ejercicios.

- **Comparación de 4 algoritmos de Boosting**: Excelente que pruebes AdaBoost, GradientBoosting, XGBoost y LightGBM. Esto demuestra conocimiento de las diferentes implementaciones.

- **Predicciones en train y test**: Correcto calcular métricas en ambos conjuntos para detectar overfitting.

- **DataFrame comparativo**: Muy clara la presentación de resultados con un DataFrame que muestra todos los modelos.

- **Conclusiones reflexivas**: Demuestras comprensión del problema al identificar el overfitting en XGBoost y LightGBM (100% en train). Tu análisis es correcto.

- **Uso de warnings.filterwarnings**: Buena práctica para limpiar la salida.

---

### 1. LEVE: Falta guardar el modelo (Paso 3 del enunciado)

El ejercicio pide explícitamente guardar el modelo en la carpeta correspondiente:

```python
# Falta esto:
import pickle

# Guardar el mejor modelo (ej: AdaBoost que tiene 79% train y 75% test, más equilibrado)
with open("../models/boosting_diabetes_model.pkl", "wb") as f:
    pickle.dump(ada, f)  # O el modelo que consideres mejor

# O si quieres guardar todos:
modelos = {
    "adaboost": ada,
    "gradient_boosting": gb,
    "xgboost": xgb,
    "lightgbm": lgb
}
with open("../models/boosting_models.pkl", "wb") as f:
    pickle.dump(modelos, f)
```

---

### 2. Falta gráfico del impacto de hiperparámetros

El enunciado pide: *"analiza su impacto con la precisión final y grafica las conclusiones"*

```python
# Ejemplo: Impacto del learning_rate en AdaBoost
learning_rates = [0.01, 0.05, 0.1, 0.5, 1.0]
train_scores = []
test_scores = []

for lr in learning_rates:
    model = AdaBoostClassifier(n_estimators=200, learning_rate=lr, random_state=42)
    model.fit(X_train, y_train)
    train_scores.append(model.score(X_train, y_train))
    test_scores.append(model.score(X_test, y_test))

plt.figure(figsize=(10, 6))
plt.plot(learning_rates, train_scores, marker='o', label='Train')
plt.plot(learning_rates, test_scores, marker='o', label='Test')
plt.xlabel('Learning Rate')
plt.ylabel('Accuracy')
plt.title('Impacto del Learning Rate en AdaBoost')
plt.legend()
plt.grid(True)
plt.show()
```

---

### 3. Faltan métricas completas de clasificación

Solo calculas accuracy, pero para un problema médico (diabetes) es importante ver precision, recall y f1, especialmente para la clase positiva (diabéticos):

```python
from sklearn.metrics import classification_report

classification_report(y_test, ada_y_pred_test, target_names=['No Diabetes', 'Diabetes'])
classification_report(y_test, gb_y_pred_test, target_names=['No Diabetes', 'Diabetes'])

# etc.
```

---

### 4. ¿Matriz de confusión para mejorarlo? :)

Para analizar qué clase se predice mejor (como pide el Paso 4):

```python
from sklearn.metrics import ConfusionMatrixDisplay

fig, axes = plt.subplots(2, 2, figsize=(12, 10))

modelos_pred = [
    (ada, ada_y_pred_test, "AdaBoost"),
    (gb, gb_y_pred_test, "Gradient Boosting"),
    (xgb, xgb_y_pred_test, "XGBoost"),
    (lgb, lgb_y_pred_test, "LightGBM")
]

for ax, (model, y_pred, name) in zip(axes.flat, modelos_pred):
    ConfusionMatrixDisplay.from_predictions(y_test, y_pred, ax=ax, display_labels=['No Diabetes', 'Diabetes'])
    ax.set_title(name)

plt.tight_layout()
plt.show()
```

---

**Comment by elius123ef** [comment]
Observaciones implementadas, cierro issue. Muchas gracias :)
